js/core/telemetrySyncManager.js

```diff
@@ -13,56 +13,83 @@ var TelemetrySyncManager = {
         var instance = TelemetrySyncManager;
         instance._teleData.push(Object.assign({}, telemetryEvent));
         if ((telemetryEvent.eid.toUpperCase() === "END") || (instance._teleData.length >= Telemetry.config.batchsize)) {
             TelemetrySyncManager.syncEvents();
         }
     },
-    updateEventStack: function(events) {
-        TelemetrySyncManager._teleData = TelemetrySyncManager._teleData.concat(events);
-    },
-    syncEvents: function() {
+    syncEvents: function(telemetryObj) {
         var Telemetry = EkTelemetry || Telemetry;
         var instance = TelemetrySyncManager;
-        var telemetryData = instance._teleData.splice(0, Telemetry.config.batchsize);
-        var telemetryObj = {
-            "id": "ekstep.telemetry",
-            "ver": Telemetry._version,
-            "ets": (new Date()).getTime(),
-            "events": telemetryData
-        };
+        if(!telemetryObj){
+            var telemetryEvents = instance._teleData.splice(0, Telemetry.config.batchsize);
+            if(!telemetryEvents.length){
+                return;
+            }
+            telemetryObj = {
+                "id": "ekstep.telemetry",
+                "ver": Telemetry._version,
+                "params": {
+                    "msgid": CryptoJS.MD5(JSON.stringify(telemetryEvents)).toString(),
+                },
+                "ets": (new Date()).getTime(),
+                "events": telemetryEvents
+            };
+        }
         var headersParam = {};
         if ('undefined' != typeof Telemetry.config.authtoken)
             headersParam["Authorization"] = 'Bearer ' + Telemetry.config.authtoken;
-
         var fullPath = Telemetry.config.host + Telemetry.config.apislug + Telemetry.config.endpoint;
         headersParam['dataType'] = 'json';
         headersParam["Content-Type"] = "application/json";
+        headersParam['x-app-id'] = Telemetry.config.pdata.id;
+        headersParam['x-device-id'] = Telemetry.fingerPrintId;
+        headersParam['x-channel-id'] = Telemetry.config.channel;
         jQuery.ajax({
             url: fullPath,
             type: "POST",
             headers: headersParam,
             data: JSON.stringify(telemetryObj)
         }).done(function(resp) {
-            console.log("Telemetry API success", resp);
+            Telemetry.config.telemetryDebugEnabled && console.log("Telemetry API success", resp);
         }).fail(function(error, textStatus, errorThrown) {
-            instance.updateEventStack(telemetryData);
+            if(instance._failedBatchSize > instance._failedBatch.length){
+                instance._failedBatch.push(telemetryObj);
+            }
             if (error.status == 403) {
-                console.error("Authentication error: ", error);
+                Telemetry.config.telemetryDebugEnabled && console.error("Authentication error: ", error);
```

Comment:
Let the errors be in console so users can report any issues. These are not debug statements